rfcs/0182-user-management.md

```diff
@@ -0,0 +1,350 @@
+  disadvantages about size and evaluation time.
+
+  Further, outside the scope of Nixpkgs, there are initiatives focused on
+  optimizing the Nix evaluators.
+
+# Alternatives
+[alternatives]: #alternatives
+
+## The trivial "do nothing"
+
+Trivially keeping the status quo.
+
+## Promote `guix-home`
+
+What? Lisp is cool!
+
+## Promote `home-manager`
+
+Now a serious alternative worthy of consideration.
+
+Home Manager has the most obvious and the most powerful advantage:
+
+It exists and is working well. Further, it is battle-tested and encodes a lot of
+knowledge in its 8 years lifespan and more than 3.7k commits.
+
+However, it has some non-negligible disadvantages:
+
+- The current code does not follow the same standards of the current Nixpkgs
+  monorepo
+
+  - many modules rely on stringly `extraConfig` instead of structured
+    `settings`, contra RFC 0042;
+
+- Many instances of technical debt, something expected in such a long-lived
+    project:
+
+    - excessive uses of `with lib;`
+
+    - hacky solutions like the dangerous `mkOutOfStoreSymlink`:
+      https://github.com/nix-community/home-manager/issues/3032
+
+    - under-documented workarounds
+
+    - too much reliance on Bash and its nasty idiosyncrasies
+
+On the other hand, starting from a cleaner slate has the opposite set of
+advantages and disadvantages: the kickstart of a project is too wide and it is
+easier to get lost. On the other hand, there is considerably more freedom to
+prototype and test ideas, older and newer.
+
+# Prior art
+[prior-art]: #prior-art
+
+## Guix Home
+
+As an example of prior art, there is our Scheme-based cousin, Guix Software
+Distribution. Since at least 2022 AD they bring a similar tool, conveniently
+called Guix Home.
+
+The nicest thing about this tool is its tight integration with Guix as a whole,
+to the point of `home` being a mere subcommand of `guix`.
+
+## Home Manager
+
+Home Manager is a well-established toolset that leverages the Nixpkgs module
+system to allow basic users to manage their user-specific environment in a
+declarative way.
+
+## Wrapper Manager
+
+Wrapper-manager is a Nix library that allows the user to configure your favorite
+applications without adding files into `~/.config`. This is done by creating
+wrapper scripts that set the appropriate environment set - variables, flags,
+configuration files etc. - to the wrapped program.
+
+## nixos-rebuild
+
+In the Nix world, nixos-rebuild is the master tool in terms of deployment of a
+Nix-based configuration, albeit being system-wide.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+Given that this tool will live inside Nixpkgs monorepo, it is expected that
+future packages will interact with this new tool. How those interactions should
+be dealt?
+
+# Future work
+[future]: #future-work
+
+- Update and extend the CI
+- Set expectations on portability
+  - Especially outside NixOS
+  - Especially outside Linux
```

Comment:
I think both of these points are going to become a real burden for future use, and I'd honestly somewhat discourage their practical application. home-manager already somewhat suffers from this "one shoe fits all" philosophy by duplicating NixOS module logic all over. Of course, for home-manager specifically this is not a problem, because they never aimed to be a NixOS only tool anyways, but if we are going to bring in first class support, we have to decide to what extent are we willing to duplicate and re-use modules from NixOS if we want to consider support for a broader ecosystem. It's not like NixOS modules can be used outside of NixOS (future RFC pending?) and the guarantees we can make in an environment outside of the module system is slim at best. It would also be practically unfeasible to attempt what home-manager does and reinvent the wheel everywhere, because that leads to duplicated logic and maintenance burden.

To be completely fair, I actually would love the idea of a world where I could declare any Linux system OOTB using Nix/NixOS modules and make it work, and optionally declare a home environment in much the same way, but practically those are hard to implement, especially at our level. The most I think we should ever realistically aim for is first-class support for NixOS only. Tighter integration allows us more flexibility in what we can re-use, more testing and applicable user experience to improve the tooling from, and although not necessarily the strongest point: makes NixOS a more appealing distribution for reasons beyond just system level configuration.

Reply:
If support for the non‐NixOS Home Manager platforms is dropped, then this will not result in merging Home Manager into the Nixpkgs monorepo or usurping Home Manager by something better; it will just result in an ecosystem fracturing between the home management tooling in the Nixpkgs monorepo and the Home Manager that will continue to exist. It is very popular to use Home Manager to manage user environments on non‐NixOS Linux and macOS, and a significant new user funnel for NixOS – besides which, the utility of Home Manager is decreased if it has to be integrated into a monolithic NixOS configuration and doesn’t allow individual users autonomy anyway, as it makes the multi‐user case a pain, and the single‐user case is already served fairly decently by NixOS already.